sd_hf_serve/README.md

```diff
@@ -0,0 +1,191 @@
+![alt text](./sdhfserve.png)
+
+For reliability, we use Karpenter. Karpenter node-pool manages Inferentia node lifecycle. Karpenter adds Inferentia nodes to handle unschedulable pods, schedules pods on those nodes, and removes them when not needed. 
+
+Our [NodePool](inf2-nodepool.yaml) uses the instance-familiy key to specify the instances needed i.e., `inf2.xlarge`, `inf2.8xlarge`, `inf2.24xlarge`, and `inf2.48xlarge`. 
+
+```yaml
+       - key: karpenter.k8s.aws/instance-family
+          operator: In
+          values: ["inf2"]
+```
+
+The Pod resource limits will determine which Inf2 instance will be requested by the total number of required neuron devices. In the example below the scheduler will launch inf2.xlarge when the replica-set size is 1 and increase linearly by the total number of needed neuron devices. 
+
+```yaml
+        resources:
+          limits:
+            aws.amazon.com/neuron: 1
+```
+
+The neuron-device-plugin returns the list of Neuron cores/devices to kublet. Kubelet advertises the Core/Device list to K8s API server (in turn to Kube-Scheduler). Kube-Scheduler picks up the pod creation request. Kube-scheduler calls the neuron-scheduler-extn filter function with a list of nodes and pod specifications. neuron-scheduler scans the nodes and filters out nodes with non contiguous cores/devices and returns the nodes that are capable of supporting the pod specification. Kube-scheduler invokes the neuron-scheduler bind function with pod and node. It updates the pod annotation with allocated neuron contiguous core/device Ids. Then neuron-scheduler sends the bind request to kubelet of the selected node. Kubelet calls the Alloc function of the neuron-device-plugin. neuron-device-plugin queries the pod annotation for allocated core/device Ids and exports the device cores to container runtime. 
+
+The solution performance comprises of the time it takes to bring neuron devices online and prepare them for process inference requests. Neuron runtime requires the model to be compiled to Neuron Executable File Format (NEFF) so that it can be executed on the Inferentia chips. Compiling a model can weigh several gigabytes and delay container startup. Therefore, we load and extract the model files from external storage to `/model` with InitContainer and mount it to the serving pod under `/app` that uses the same `workday` volume.
+
+```yaml
+     initContainers:
+      - name: pull-model-tar
+        image: ${AWS_ACCOUNT_ID}.dkr.ecr.${AWS_REGION}.amazonaws.com/stablediffusion:amd64-neuron-assets
+        imagePullPolicy: Always
+        volumeMounts:
+        - name: workdir
+          mountPath: /model
+        command:
+        - /bin/bash
+        - "-exc"
+        - |
+          set -x
+          aws s3 cp s3://sdinfer/stable-diffusion-2-1-base.tar.gz /model/model.tar.gz
```

Comment:
Is this a public S3?

Reply:
No see https://github.com/yahavb/aws-neuron-eks-samples/blob/a171acb68a71acee5cbe7161fccd6cce4809decf/sd_hf_serve/oci-image-build/pipeline-stack.ts#L36